### PR TITLE
Fix for program crash when destructor is called before channel close with blocked readers/writers

### PR DESCRIPTION
**Pull request by abhinavarora**
Fix https://github.com/PaddlePaddle/Paddle/issues/8248

### CODE REVIEW

File: paddle/framework/details/buffered_channel.h [hunk 1]
```diff
@@ -67,20 +71,30 @@ bool Buffered<T>::Send(T* item) {
...
+  send_ctr--;
+  destructor_cond_var_.notify_one();
   return ret;
 }
 
 template <typename T>
 bool Buffered<T>::Receive(T* item) {
+  bool ret = false;
+  // Once the channel has been closed and all data has been consumed,
+  // just return false. Don't even try acquiring the mutex.
```
Comment (abhinavarora) [inline paddle/framework/details/buffered_channel.h]: If a reader thread is allowed to acquire mutex here, then in a hypothetical infinite readers scenario, the destructor could be blocked forever. So after closing and when all the data in the channel is consumed, any call to `Receive` should immediately return without even trying to acquire the lock.

File: paddle/framework/channel_test.cc [hunk 2]
```diff
@@ -60,13 +60,83 @@ TEST(Channel, SufficientBufferSizeDoesntBlock) {
...
+  // also panic because the queue is empty.
+  CloseChannel(ch);
+
+  send_thread = std::thread{[&]() {
+    size_t i = data;
+    EXPECT_EQ(ch->Send(&i), false);  // should panic because channel is closed
+  }};
+  recv_thread = std::thread{[&]() {
+    size_t i;
+    // should panic because channel is closed and queue is empty
```
Comment (kavyasrinet) [inline paddle/framework/channel_test.cc]: Same as above.

File: paddle/framework/channel_test.cc [hunk 3]
```diff
@@ -60,13 +60,83 @@ TEST(Channel, SufficientBufferSizeDoesntBlock) {
...
+    EXPECT_EQ(ch->Receive(&i), false);
+  }};
+
+  send_thread.join();
+  recv_thread.join();
```
Comment (kavyasrinet) [inline paddle/framework/channel_test.cc]: Also, I think the function `ClosedUnBufferedChannelPanics` and the function `ClosedBufferedChannelPanics` reuse a block of code as a part of the functionality.
I propose that we write the reused block inside another function and call that function with a buffered and unbuffered channel , check PR: https://github.com/PaddlePaddle/Paddle/pull/8162 for reference.

File: paddle/framework/channel_test.cc [hunk 4]
```diff
@@ -381,3 +451,163 @@ TEST(Channel, UnbufferedMoreReceiveLessSendTest) {
...
+  for (size_t i = 0; i < num_threads; i++) {
+    EXPECT_EQ(thread_ended[i], true);
+  }
+
+  for (size_t i = 0; i < num_threads; i++) t[i].join();
+}
+
+// This tests that destroying an unbuffered channel also unblocks
+//  unblocks any senders waiting for senders
+TEST(Channel, UnbufferedChannelDestroyUnblocksSendersTest) {
```
Comment (kavyasrinet) [inline paddle/framework/channel_test.cc]: Same comment as above, I think there is a lot of repetition of code inside the functions of Unbuffered and Buffered functions. We can write a helper function that implements the block and call it with different channels.

File: paddle/framework/details/buffered_channel.h [hunk 5]
```diff
@@ -100,6 +114,12 @@ Buffered<T>::~Buffered() {
   closed_ = true;
   channel_.clear();
   NotifyAllParticipants(&lock);
+
+  // The destructor must wait for all readers and writers to complete their task
+  // The channel has been closed, so we will not accept new readers and writers
+  lock.lock();
+  destructor_cond_var_.wait(
```
Comment (kavyasrinet) [inline paddle/framework/details/buffered_channel.h]: Thanks for catching this!

File: paddle/framework/channel_test.cc (hunk 3 above)
Comment (chengduoZH) [reply paddle/framework/channel_test.cc]: I agreed with @kavyasrinet .

File: paddle/framework/channel_test.cc [hunk 6]
```diff
@@ -60,13 +60,52 @@ TEST(Channel, SufficientBufferSizeDoesntBlock) {
...
 
-TEST(Channel, SendOnClosedChannelPanics) {
-  const size_t buffer_size = 10;
-  auto ch = MakeChannel<size_t>(buffer_size);
-  size_t i = 5;
-  EXPECT_EQ(ch->Send(&i), true);  // should not block or panic
+// This tests that a  channel must return false
+// on send and receive performed after closing the channel.
+// Receive will only return false after close when queue is empty
+void ClosedChannelSendReceivePanics(Channel<size_t> *ch) {
```
Comment (wangkuiyi) [inline paddle/framework/channel_test.cc]: Is `SendReceiveWithACloseChannelShouldPanic` a more accurate name for this test?

File: paddle/framework/channel_test.cc [hunk 7]
```diff
@@ -60,13 +60,52 @@ TEST(Channel, SufficientBufferSizeDoesntBlock) {
...
+  const size_t data = 5;
+  std::thread send_thread{[&]() {
+    size_t i = data;
+    EXPECT_EQ(ch->Send(&i), true);  // should not block
```
Comment (wangkuiyi) [inline paddle/framework/channel_test.cc]: Can we remove the definition of the variable `i` and call `ch-Send(&data)` here?

File: paddle/framework/channel_test.cc [hunk 8]
```diff
@@ -60,13 +60,52 @@ TEST(Channel, SufficientBufferSizeDoesntBlock) {
...
-  const size_t buffer_size = 10;
-  auto ch = MakeChannel<size_t>(buffer_size);
-  size_t i = 5;
-  EXPECT_EQ(ch->Send(&i), true);  // should not block or panic
+// This tests that a  channel must return false
+// on send and receive performed after closing the channel.
+// Receive will only return false after close when queue is empty
+void ClosedChannelSendReceivePanics(Channel<size_t> *ch) {
+  const size_t data = 5;
+  std::thread send_thread{[&]() {
```
Comment (wangkuiyi) [inline paddle/framework/channel_test.cc]: I think we need a comment here:

```c++
// By creating separate threads for sending and receiving, we make this 
// function able to test both buffered and unbuffered channels.
```

File: paddle/framework/channel_test.cc [hunk 9]
```diff
@@ -60,13 +60,83 @@ TEST(Channel, SufficientBufferSizeDoesntBlock) {
...
+  // After closing send should panic. Receive should
+  // also panic because the queue is empty.
   CloseChannel(ch);
-  EXPECT_EQ(ch->Send(&i), false);  // should panic
+
+  send_thread = std::thread{[&]() {
+    size_t i = data;
+    EXPECT_EQ(ch->Send(&i), false);  // should panic because channel is closed
+  }};
+  recv_thread = std::thread{[&]() {
```
[the remark this answers is not in the document]
Comment (abhinavarora) [reply paddle/framework/channel_test.cc]: Thank you fixed that in the latest commit.

File: paddle/framework/channel_test.cc (hunk 2 above)
Comment (abhinavarora) [reply paddle/framework/channel_test.cc]: Fixed in the latest commit

File: paddle/framework/channel_test.cc (hunk 3 above)
Comment (abhinavarora) [reply paddle/framework/channel_test.cc]: Fixed in the latest commit. Thank you

File: paddle/framework/channel_test.cc (hunk 4 above)
Comment (abhinavarora) [reply paddle/framework/channel_test.cc]: Fixed in the latest commit. Thank you.

File: paddle/framework/channel_test.cc (hunk 6 above)
Comment (abhinavarora) [reply paddle/framework/channel_test.cc]: Fixed. Thank you.

File: paddle/framework/channel_test.cc (hunk 8 above)
Comment (abhinavarora) [reply paddle/framework/channel_test.cc]: Done.

**Pull request merged**